src/wh_server_crypto.c

```diff
@@ -649,6 +647,79 @@ int wh_Server_HandleCryptoRequest(whServerContext* server,
+        /* in, out and key are after the fixed size fields */
+        in = (uint8_t*)(&packet->cmacReq + 1);
+        key = in + packet->cmacReq.inSz;
+        out = (uint8_t*)(&packet->cmacRes + 1);
+        /* do oneshot if all fields are present */
+        if (packet->cmacReq.inSz != 0 && packet->cmacReq.keySz != 0 &&
+            packet->cmacReq.outSz != 0) {
+            field = packet->cmacReq.outSz;
+            ret = wc_AesCmacGenerate_ex(server->crypto->cmac, out, &field, in,
+                packet->cmacReq.inSz, key, packet->cmacReq.keySz, NULL,
+                server->crypto->devId);
+            packet->cmacRes.outSz = field;
+        }
+        else {
+            /* do each operation based on which fields are set */
+            /* set key if present, otherwise load the struct from keyId */
+            if (packet->cmacReq.keySz != 0) {
+                /* initialize cmac with key and type */
+                ret = wc_InitCmac_ex(server->crypto->cmac, key,
+                    packet->cmacReq.keySz, packet->cmacReq.type, NULL, NULL,
+                    server->crypto->devId);
+            }
+            else {
+                field = sizeof(server->crypto->cmac);
+                ret = hsmReadKey(server,
+                    MAKE_WOLFHSM_KEYID(WOLFHSM_KEYTYPE_CRYPTO,
+                    server->comm->client_id, packet->cmacReq.keyId), NULL,
+                    (uint8_t*)server->crypto->cmac, &field);
+            }
+            if (ret == 0 && packet->cmacReq.inSz != 0) {
+                ret = wc_CmacUpdate(server->crypto->cmac, in,
+                    packet->cmacReq.inSz);
+            }
+            /* do final and evict the struct if outSz is set, otherwise cache
+             * the struct for a future call */
+            if (ret == 0 && packet->cmacReq.outSz != 0) {
+                keyId = packet->cmacReq.keyId;
+                field = packet->cmacReq.outSz;
+                ret = wc_CmacFinal(server->crypto->cmac, out, &field);
+                packet->cmacReq.outSz = field;
+                /* evict the key */
+                if (ret == 0) {
+                    hsmEvictKey(server,
+                        MAKE_WOLFHSM_KEYID(WOLFHSM_KEYTYPE_CRYPTO,
+                        server->comm->client_id, keyId));
+                }
+            }
+            else {
+                /* cache/re-cache updated struct */
+                XMEMSET((uint8_t*)meta, 0, sizeof(meta));
+                if (packet->cmacReq.keySz != 0) {
+                    keyId = WOLFHSM_KEYTYPE_CRYPTO;
+                    ret = hsmGetUniqueId(server, &keyId);
+                }
+                else {
+                    keyId = MAKE_WOLFHSM_KEYID(WOLFHSM_KEYTYPE_CRYPTO,
+                        server->comm->client_id, packet->cmacReq.keyId);
+                }
+                meta->id = keyId;
+                meta->len = sizeof(server->crypto->cmac);
+                ret = hsmCacheKey(server, meta, (uint8_t*)server->crypto->cmac);
```

Comment:
So this is storing the entire CMAC context as a CacheKey?  Hmm... Gonna have to stew on this if so.